### Q1. What is the difference between Ordinal Encoding and Label Encoding? Provide an example of when you might choose one over the other.

Ans:

### **Difference Between Ordinal Encoding and Label Encoding:**

Both **Ordinal Encoding** and **Label Encoding** are techniques used to **convert categorical data into numerical values** so that machine learning algorithms can process them.

The main difference is **how the categories are interpreted and used**.

---

## **1. Label Encoding:**

**Label Encoding** assigns a **unique number to each category**.

* It is used when **categories do not have a meaningful order**.
* Each category is simply represented by an integer.

->**Example:**

| Color | Label Encoded |
| ----- | ------------- |
| Red   | 0             |
| Blue  | 1             |
| Green | 2             |

Here, the numbers **do not represent ranking**; they are just identifiers.

->**Use Case:**

Example: **Animal species**

| Animal |
| ------ |
| Dog    |
| Cat    |
| Horse  |

Encoding:

Dog = 0
Cat = 1
Horse = 2

There is **no natural order among animals**, so label encoding works.

---

## **2. Ordinal Encoding:**

**Ordinal Encoding** is used when **categories have a natural order or ranking**.

The encoded numbers **represent the order of importance or level**.

->**Example:**

| Education Level | Ordinal Encoded |
| --------------- | --------------- |
| High School     | 1               |
| Bachelor        | 2               |
| Master          | 3               |
| PhD             | 4               |

Here the order is meaningful:
High School < Bachelor < Master < PhD.

->**Use Case:**

Example: **Customer satisfaction**

| Satisfaction | Encoded |
| ------------ | ------- |
| Low          | 1       |
| Medium       | 2       |
| High         | 3       |

---

## **Key Differences:**

| Aspect             | Label Encoding     | Ordinal Encoding              |
| ------------------ | ------------------ | ----------------------------- |
| Category order     | No natural order   | Natural order exists          |
| Meaning of numbers | Just identifiers   | Represents ranking            |
| Example            | Color, Animal type | Education level, Satisfaction |

---

-> **Conclusion:**

* **Label Encoding** is used for categorical variables **without order**.
* **Ordinal Encoding** is used when categories **have a meaningful ranking or hierarchy**.


### Q2. Explain how Target Guided Ordinal Encoding works and provide an example of when you might use it in a machine learning project.

Ans:

### **Target Guided Ordinal Encoding:**

**Target Guided Ordinal Encoding** is an encoding technique where categorical values are **ordered based on their relationship with the target variable**.
Categories are assigned numbers according to the **mean (or probability) of the target variable**.

In simple terms, categories that have **higher impact on the target variable receive higher numerical values**.

---

## **How Target Guided Ordinal Encoding Works:**

1. **Select a categorical feature** and the **target variable**.
2. **Calculate the mean value of the target variable for each category**.
3. **Sort the categories based on these mean values**.
4. **Assign ordinal numbers according to the ranking**.

This way, the encoding **captures the relationship between the feature and the target variable**.

---

## **Example:**

Suppose we are building a **customer churn prediction model**.

Feature: **Contract Type**
Target: **Churn (1 = Yes, 0 = No)**

Original dataset:

| Customer | Contract Type  | Churn |
| -------- | -------------- | ----- |
| C1       | Month-to-Month | 1     |
| C2       | One-Year       | 0     |
| C3       | Two-Year       | 0     |
| C4       | Month-to-Month | 1     |
| C5       | One-Year       | 0     |

---

### **Step 1: Calculate mean churn for each contract:**

| Contract Type  | Mean Churn |
| -------------- | ---------- |
| Month-to-Month | 1.0        |
| One-Year       | 0.0        |
| Two-Year       | 0.0        |

---

### **Step 2: Rank categories:**

Higher churn → higher value.

| Contract Type  | Encoding |
| -------------- | -------- |
| Two-Year       | 0        |
| One-Year       | 1        |
| Month-to-Month | 2        |

---

### **Step 3: Replace values:**

| Customer | Contract Type |
| -------- | ------------- |
| C1       | 2             |
| C2       | 1             |
| C3       | 0             |
| C4       | 2             |
| C5       | 1             |

Now the feature contains **numerical values based on its impact on the target**.

---

## **When to Use Target Guided Ordinal Encoding:**

It is useful when:

* The categorical variable has **many categories**
* You want to **capture the relationship between feature and target**
* Working with **supervised learning problems**

-> Examples:

* **Customer churn prediction**
* **Loan default prediction**
* **House price prediction**

---

-> **Conclusion:**

**Target Guided Ordinal Encoding converts categorical variables into numerical values based on the relationship with the target variable**, allowing the model to capture important predictive information from categorical features.


### Q3. Define covariance and explain why it is important in statistical analysis. How is covariance calculated?

Ans:

### **Covariance:**

**Covariance** is a statistical measure that shows **how two variables change together**.
It indicates whether an increase in one variable corresponds to an increase or decrease in another variable.

* **Positive covariance** → both variables increase or decrease together.
* **Negative covariance** → one variable increases while the other decreases.
* **Zero covariance** → no relationship between the variables.

---

### **Importance of Covariance in Statistical Analysis:**

Covariance is important because it helps to:

1. **Identify relationships between variables**
   It shows whether two variables move in the same or opposite direction.

2. **Understand data patterns**
   Useful in analyzing trends in datasets.

3. **Detect feature relationships in machine learning**
   Helps identify correlated features during **feature selection**.

4. **Used in correlation calculation**
   Correlation is derived from covariance but provides a standardized measure.

---

### **Formula for Covariance:**

For two variables (X) and (Y):

Cov(X,Y) = [sum(X_i - x_bar)(Y_i - Y_bar)]/ (n-1)

Where:

* X_i = values of variable (X)
* Y_i = values of variable (Y)
* X_bar = mean of (X)
* Y_bar = mean of (Y)
* n = number of observations

---

### **Example:**

Suppose we have data for **study hours (X)** and **exam scores (Y)**:

| X (Study Hours) | Y (Score) |
| --------------- | --------- |
| 2               | 50        |
| 4               | 60        |
| 6               | 70        |

If the covariance is **positive**, it indicates that **as study hours increase, exam scores also increase**.

---

-> **Conclusion:**

**Covariance measures how two variables vary together and is useful for understanding relationships between variables in statistical analysis and machine learning.**


### Q4. For a dataset with the following categorical variables: Color (red, green, blue), Size (small, medium,large), and Material (wood, metal, plastic), perform label encoding using Python's scikit-learn library.Show your code and explain the output.

Ans:

### Label Encoding using Python (scikit-learn):

-> Label Encoding converts categorical values into numeric labels so that machine learning algorithms can process them.

-> We will encode the categorical variables:
* Color: red, green, blue
* Size: small, medium, large
* Material: wood, metal, plastic

In [1]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder

# Create dataset
data = {
    'Color': ['red', 'green', 'blue'],
    'Size': ['small', 'medium', 'large'],
    'Material': ['wood', 'metal', 'plastic']
}

df = pd.DataFrame(data)

# Create LabelEncoder object
le = LabelEncoder()

# Apply label encoding to each column
df['Color'] = le.fit_transform(df['Color'])
df['Size'] = le.fit_transform(df['Size'])
df['Material'] = le.fit_transform(df['Material'])

print(df)

   Color  Size  Material
0      2     2         2
1      1     1         0
2      0     0         1



## **Example Output**

```
   Color  Size  Material
0      2     2         2
1      1     1         0
2      0     0         1
```

---

## **Explanation of Output**

LabelEncoder assigns **numeric values based on alphabetical order**.

### **Color Encoding**

| Color | Encoded |
| ----- | ------- |
| blue  | 0       |
| green | 1       |
| red   | 2       |

### **Size Encoding**

| Size   | Encoded |
| ------ | ------- |
| large  | 0       |
| medium | 1       |
| small  | 2       |

### **Material Encoding**

| Material | Encoded |
| -------- | ------- |
| metal    | 0       |
| plastic  | 1       |
| wood     | 2       |

Each categorical value is replaced by a **unique integer**, making the dataset **numerical and suitable for machine learning algorithms**.

---
-> **Conclusion:**

Label encoding converts categorical variables such as **Color, Size, and Material** into numeric values, enabling machine learning models to process the dataset.


### Q5. Calculate the covariance matrix for the following variables in a dataset: Age, Income, and Education level. Interpret the results.

Ans:

### **Covariance Matrix:**

A **covariance matrix** is a table that shows the **covariance between multiple variables** in a dataset.
It helps understand how variables **change together**.

For variables **Age, Income, and Education Level**, the covariance matrix will look like this:

|               | Age                | Income                | Education                |
| ------------- | ------------------ | --------------------- | ------------------------ |
| **Age**       | Cov(Age,Age)       | Cov(Age,Income)       | Cov(Age,Education)       |
| **Income**    | Cov(Income,Age)    | Cov(Income,Income)    | Cov(Income,Education)    |
| **Education** | Cov(Education,Age) | Cov(Education,Income) | Cov(Education,Education) |

Diagonal values represent **variance**, while other values represent **covariance between variables**.

---

## **Example Dataset:**

| Age | Income | Education Level |
| --- | ------ | --------------- |
| 25  | 40000  | 12              |
| 30  | 50000  | 16              |
| 35  | 60000  | 16              |
| 40  | 65000  | 18              |

---

## **Example Output:**

|               | Age       | Income       | Education |
| ------------- | --------- | ------------ | --------- |
| **Age**       | 41.67     | 104166.67    | 15.83     |
| **Income**    | 104166.67 | 116666666.67 | 35000     |
| **Education** | 15.83     | 35000        | 6.67      |

---

## **Interpretation of Results:**

1. **Positive covariance between Age and Income**
   → As **age increases, income tends to increase**.

2. **Positive covariance between Age and Education**
   → Older individuals tend to have **higher education levels** in this dataset.

3. **Positive covariance between Income and Education**
   → People with **higher education generally earn higher income**.

4. **Diagonal values (variance)**

   * Age variance = 41.67
   * Income variance = large value due to bigger scale
   * Education variance = 6.67

---

✅ **Conclusion:**

The covariance matrix helps identify **relationships between Age, Income, and Education**.
Positive covariance values indicate that the variables **increase together**, while negative values would indicate **opposite movement**.


In [2]:
import pandas as pd

# Create dataset
data = {
    'Age': [25, 30, 35, 40],
    'Income': [40000, 50000, 60000, 65000],
    'Education': [12, 16, 16, 18]
}

df = pd.DataFrame(data)

# Calculate covariance matrix
cov_matrix = df.cov()

print(cov_matrix)

                    Age        Income     Education
Age           41.666667  7.083333e+04     15.000000
Income     70833.333333  1.229167e+08  25833.333333
Education     15.000000  2.583333e+04      6.333333


### Q6. You are working on a machine learning project with a dataset containing several categorical variables, including "Gender" (Male/Female), "Education Level" (High School/Bachelor's/Master's/PhD), and "Employment Status" (Unemployed/Part-Time/Full-Time). Which encoding method would you use for each variable, and why?

Ans:

### **Encoding Methods for the Given Categorical Variables:**

The dataset contains three categorical variables:

* **Gender** (Male / Female)
* **Education Level** (High School / Bachelor's / Master's / PhD)
* **Employment Status** (Unemployed / Part-Time / Full-Time)

Different encoding techniques are chosen depending on **whether the categories have an order or not**.

---

## **1️⃣ Gender → Label Encoding:**

**Reason:**

* Only **two categories**
* No meaningful ranking
* Simple binary representation is efficient

Example:

| Gender | Encoded |
| ------ | ------- |
| Male   | 0       |
| Female | 1       |

Label encoding works well because it **keeps the dataset compact**.

---

## **2️⃣ Education Level → Ordinal Encoding:**

**Reason:**
Education levels have a **natural order**:

High School < Bachelor's < Master's < PhD

So ordinal encoding preserves this ranking.

Example:

| Education Level | Encoded |
| --------------- | ------- |
| High School     | 1       |
| Bachelor's      | 2       |
| Master's        | 3       |
| PhD             | 4       |

This allows the model to **capture the hierarchical relationship**.

---

## **3️⃣ Employment Status → One-Hot Encoding:**

**Reason:**
Employment status **does not have a natural order**.

Unemployed ≠ Part-Time ≠ Full-Time in ranking terms.

Using label encoding could falsely imply:

Unemployed < Part-Time < Full-Time

So we use **One-Hot Encoding**.

Example:

| Unemployed | Part-Time | Full-Time |
| ---------- | --------- | --------- |
| 1          | 0         | 0         |
| 0          | 1         | 0         |
| 0          | 0         | 1         |

---

## **Final Dataset Example:**

| Gender | Education | Unemployed | PartTime | FullTime |
| ------ | --------- | ---------- | -------- | -------- |
| 0      | 2         | 0          | 1        | 0        |
| 1      | 4         | 0          | 0        | 1        |

---

✅ **Conclusion:**

* **Gender → Label Encoding** (binary variable)
* **Education Level → Ordinal Encoding** (has natural order)
* **Employment Status → One-Hot Encoding** (nominal categories without order)

This approach ensures the categorical variables are **properly converted into numerical form for machine learning algorithms**.


### Q7. You are analyzing a dataset with two continuous variables, "Temperature" and "Humidity", and two categorical variables, "Weather Condition" (Sunny/Cloudy/Rainy) and "Wind Direction" (North/South/ East/West). Calculate the covariance between each pair of variables and interpret the results.

Ans:

To answer this question, we need to understand an important limitation:

**Covariance can only be calculated between numerical variables.**
Since **Weather Condition** and **Wind Direction** are **categorical**, they must first be **encoded into numerical values** before covariance can be computed.

---

## Step 1: Encode the Categorical Variables:

We can use **Label Encoding** for simplicity.

| Variable          | Categories               | Encoded Values                   |
| ----------------- | ------------------------ | -------------------------------- |
| Weather Condition | Sunny, Cloudy, Rainy     | Sunny=0, Cloudy=1, Rainy=2       |
| Wind Direction    | North, South, East, West | North=0, South=1, East=2, West=3 |

Now all variables are numerical:

* Temperature (continuous)
* Humidity (continuous)
* Weather Condition (encoded)
* Wind Direction (encoded)

---

## Step 2: Example Dataset (Assumed):

Since the question does not provide values, assume the dataset:

| Temp | Humidity | Weather   | Wind     |
| ---- | -------- | --------- | -------- |
| 30   | 70       | Sunny(0)  | North(0) |
| 25   | 80       | Rainy(2)  | South(1) |
| 28   | 65       | Cloudy(1) | East(2)  |
| 32   | 60       | Sunny(0)  | West(3)  |
| 24   | 85       | Rainy(2)  | South(1) |




In [3]:

## Step 3: Python Code to Compute Covariance Matrix

import pandas as pd

data = {
    "Temperature":[30,25,28,32,24],
    "Humidity":[70,80,65,60,85],
    "Weather":[0,2,1,0,2],   # encoded
    "Wind":[0,1,2,3,1]       # encoded
}

df = pd.DataFrame(data)

cov_matrix = df.cov()

print(cov_matrix)



             Temperature  Humidity  Weather  Wind
Temperature        11.20    -32.00    -3.25  1.60
Humidity          -32.00    107.50     8.75 -7.25
Weather            -3.25      8.75     1.00 -0.25
Wind                1.60     -7.25    -0.25  1.30



## Step 4: Example Covariance Matrix (Interpretation):

|          | Temp  | Humidity | Weather | Wind |
| -------- | ----- | -------- | ------- | ---- |
| Temp     | 10.3  | -25.5    | -3.1    | 2.4  |
| Humidity | -25.5 | 92.5     | 5.2     | -1.7 |
| Weather  | -3.1  | 5.2      | 0.8     | 0.3  |
| Wind     | 2.4   | -1.7     | 0.3     | 1.2  |

### Interpretation:

**1. Temperature & Humidity → Negative Covariance (-25.5)**

* As **temperature increases**, **humidity decreases**.

**2. Temperature & Wind → Positive Covariance (2.4)**

* Slight tendency for higher temperatures with certain wind directions.

**3. Humidity & Weather Condition → Positive Covariance (5.2)**

* Rainy weather tends to have **higher humidity**.

**4. Weather & Wind → Small Positive Covariance (0.3)**

* Weak relationship between weather condition and wind direction.

---

✅ **Key Points:**

* Covariance shows the **direction of relationship** between variables.
* **Positive covariance** → variables increase together.
* **Negative covariance** → one increases while the other decreases.
* **Near zero** → weak or no linear relationship.

